plugins/out_azure_blob/azure_blob_store.c

```diff
@@ -0,0 +1,454 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2015-2024 The Fluent Bit Authors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "azure_blob_store.h"
+#include <fluent-bit/flb_output_plugin.h>
+#include <fluent-bit/flb_fstore.h>
+#include <fluent-bit/flb_time.h>
+#include <fcntl.h>
+
+/*
+ * Simple and fast hashing algorithm to create keys in the local buffer
+ */
+static flb_sds_t gen_store_filename(const char *tag)
+{
+    int c;
+    unsigned long hash = 5381;
+    unsigned long hash2 = 5381;
+    flb_sds_t hash_str;
+    flb_sds_t tmp;
+    struct flb_time tm;
+
+    /* get current time */
+    flb_time_get(&tm);
+
+    /* compose hash */
+    while ((c = *tag++)) {
+        hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
+    }
+    hash2 = (unsigned long) hash2 * tm.tm.tv_sec * tm.tm.tv_nsec;
+
+    /* flb_sds_printf allocs if the incoming sds is not at least 64 bytes */
+    hash_str = flb_sds_create_size(64);
+    if (!hash_str) {
+        flb_errno();
+        return NULL;
+    }
+    tmp = flb_sds_printf(&hash_str, "%lu-%lu", hash, hash2);
+    if (!tmp) {
+        flb_errno();
+        flb_sds_destroy(hash_str);
```

Comment:
if a sds allocation fails flb_errno() [is already called](https://github.com/fluent/fluent-bit/blob/master/src/flb_sds.c#L44), so this one is not needed.